pkg/deployer/lib/resourcemanager/objectapplier.go

```diff
@@ -351,91 +356,84 @@ func (a *ManifestApplier) injectLabels(obj client.Object) {
 	obj.SetLabels(labels)
 }
 
-// cleanupOrphanedResources removes all managed resources that are not rendered anymore.
-func (a *ManifestApplier) cleanupOrphanedResources(ctx context.Context, managedResources []managedresource.ManagedResourceStatus) error {
-	var (
-		allErrs              []error
-		containsTimeoutError bool
-		wg                   sync.WaitGroup
-		errMux               sync.Mutex
-	)
-	logger, ctx := logging.FromContextOrNew(ctx, nil, lc.KeyMethod, "cleanupOrphanedResources")
+func (a *ManifestApplier) cleanupOrphanedResourcesInGroups(ctx context.Context, oldManagedResources []managedresource.ManagedResourceStatus) error {
+	logger, ctx := logging.FromContextOrNew(ctx, nil, lc.KeyMethod, "cleanupOrphanedResourcesInGroups")
+	orphanedManagedResources := []managedresource.ManagedResourceStatus{}
 
-	var timeoutErr lserrors.LsError
+	_, err := timeout.TimeoutExceeded(ctx, a.deployItem, TimeoutCheckpointDeployerCleanupOrphaned)
+	if err != nil {
+		return err
+	}
 
-	for _, mr := range managedResources {
-		logger2 := logger.WithValues(lc.KeyResource, types.NamespacedName{Namespace: mr.Resource.Namespace, Name: mr.Resource.Name}.String(), lc.KeyResourceKind, mr.Resource.Kind)
-		logger2.Debug("Checking resource")
-		if mr.Policy == managedresource.IgnorePolicy || mr.Policy == managedresource.KeepPolicy {
-			logger2.Debug("Ignoring resource due to policy", lc.KeyManagedResourcePolicy, string(mr.Policy))
-			continue
-		}
-		ref := mr.Resource
-		obj := kutil.ObjectFromCoreObjectReference(&ref)
-		key := kutil.ObjectKey(ref.Name, ref.Namespace)
+	for i := range oldManagedResources {
+		mr := &oldManagedResources[i]
 
-		if err := read_write_layer.GetUnstructured(ctx, a.kubeClient, key, obj, read_write_layer.R000049); err != nil {
-			if apierrors.IsNotFound(err) {
-				logger2.Debug("Object not found")
-				continue
-			}
-			return fmt.Errorf("unable to get object %s %s: %w", obj.GroupVersionKind().String(), obj.GetName(), err)
-		}
+		mrLogger, mrCtx := logger.WithValuesAndContext(ctx,
+			lc.KeyResource, types.NamespacedName{Namespace: mr.Resource.Namespace, Name: mr.Resource.Name}.String(),
+			lc.KeyResourceKind, mr.Resource.Kind)
+		mrLogger.Debug("Checking resource")
 
-		// if fallback policy is set and the resource is already managed by another deployer
-		// we are not allowed to manage that resource
-		if mr.Policy == managedresource.FallbackPolicy && !kutil.HasLabelWithValue(obj, manifestv1alpha2.ManagedDeployItemLabel, a.deployItemName) {
-			logger.Info("Resource is already managed, skip cleanup", lc.KeyResource, key.String())
+		ok, err := a.filterByPolicy(mrCtx, mr)
```

Comment:
There are two nearly identical filterByPolicy functions one method in the objectapplier and one function in the manifest ensure.go file. Are these really different?

Reply:
I have removed the duplicate method in manifest ensure.go file.